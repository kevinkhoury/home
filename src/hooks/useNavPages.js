import { useStaticQuery, graphql } from 'gatsby'

export default () => {
  const query = useStaticQuery(
    graphql`
      query NavPagesQuery {
        markdownRemark(frontmatter: { templateKey: { eq: "menu-data" } }) {
          frontmatter {
            menuItems {
              slug
              label
            }
          }
        }
      }
    `,
  )
  const { menuItems } = query.markdownRemark.frontmatter
  return menuItems.map(({ slug: orginalSlug, label }) => {
    // normalize to initial and trailing slash
    const slug =
      orginalSlug === '/'
        ? orginalSlug
        : `/${orginalSlug
            .split('/')
            .filter(x => x)
            .join('/')}/`
    return { slug, label }
  })
  // return query.markdownRemark.frontmatter.menuItems
}
