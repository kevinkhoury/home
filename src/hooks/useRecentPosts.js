import { useStaticQuery, graphql } from 'gatsby'
import { getValidDates } from '../utils'

export default () => {
  const data = useStaticQuery(
    graphql`
      query QueryRecentPosts {
        allMarkdownRemark(
          sort: { order: DESC, fields: [frontmatter___date] }
          filter: {
            frontmatter: {
              templateKey: { eq: "blog-post" }
              published: { eq: true }
            }
          }
          limit: 4
        ) {
          edges {
            node {
              fields {
                slug
                gitAuthorTime
                gitCreatedTime
              }
              frontmatter {
                date(formatString: "MMM D, YYYY")
                pageTitle
                featuredImage {
                  src {
                    childImageSharp {
                      fluid {
                        originalName
                      }
                      original {
                        height
                        width
                      }
                    }
                  }
                  m: src {
                    childImageSharp {
                      fluid(maxWidth: 500, maxHeight: 664) {
                        ...GatsbyImageSharpFluid_withWebp
                        originalName
                      }
                      original {
                        height
                        width
                      }
                    }
                  }
                  d: src {
                    childImageSharp {
                      fluid(maxWidth: 1000, maxHeight: 664) {
                        ...GatsbyImageSharpFluid_withWebp
                        originalName
                      }
                      original {
                        height
                        width
                      }
                    }
                  }
                  alt
                }
              }
            }
          }
        }
      }
    `,
  )
  if (
    data &&
    data.allMarkdownRemark &&
    data.allMarkdownRemark.edges &&
    data.allMarkdownRemark.edges.length > 0
  ) {
    return data.allMarkdownRemark.edges.map(({ node }) => {
      const {
        frontmatter: { featuredImage, pageTitle, date: userDate },
        fields: { slug, gitAuthorTime, gitCreatedTime },
      } = node
      const { date, dateModified } = getValidDates(
        userDate,
        gitAuthorTime,
        gitCreatedTime,
      )
      return {
        image: !!featuredImage ? featuredImage : null,
        slug,
        pageTitle,
        date,
        dateModified,
      }
    })
  } else {
    return []
  }
}
