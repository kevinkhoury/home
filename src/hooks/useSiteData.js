import { useStaticQuery, graphql } from 'gatsby'

export default () => {
  const query = useStaticQuery(
    graphql`
      query SiteDataQuery {
        markdownRemark(frontmatter: { templateKey: { eq: "site-data" } }) {
          frontmatter {
            name
            jobTitle
            siteName
            siteUrl
            socialLinks {
              twitter {
                url
                show
              }
              facebook {
                url
                show
              }
              linkedin {
                url
                show
              }
              pinterest {
                url
                show
              }
              instagram {
                url
                show
              }
            }
            themeOptions {
              colorScheme
              fontScheme
              showThemeSwitcher
            }
            fallbackImage {
              childImageSharp {
                original {
                  height
                  width
                }
                fluid(maxWidth: 1440) {
                  ...GatsbyImageSharpFluid_withWebp
                  originalName
                }
              }
            }
          }
        }
      }
    `,
  )
  return query.markdownRemark.frontmatter
}
