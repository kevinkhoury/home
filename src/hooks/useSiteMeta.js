import { useStaticQuery, graphql } from 'gatsby'

export default () => {
  const query = useStaticQuery(
    graphql`
      query QuerySiteMetadata {
        site {
          siteMetadata {
            title
            colorOptions {
              value
              label
            }
            fontOptions {
              value
              label
            }
          }
        }
      }
    `,
  )
  return query.site.siteMetadata
}
