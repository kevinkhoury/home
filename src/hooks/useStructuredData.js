const buildImage = (image) => ({
  '@context': 'http://schema.org/',
  '@type': 'ImageObject',
  url: image.url,
  height: image.height,
  width: image.width,
})

const personSchema = ({ name, jobTitle, sameAs }) => ({
  '@context': 'http://schema.org/',
  '@type': 'Person',
  '@id': '#person',
  name,
  jobTitle,
  sameAs,
})

const publisherSchema = ({ siteName, logo }) => ({
  '@context': 'http://schema.org/',
  '@type': 'Organization',
  '@id': '#publisher',
  name: siteName,
  logo: buildImage(logo),
})

const webpageSchema = ({
  headline,
  url,
  description,
  image,
  name,
  jobTitle,
  schemaType,
  sameAs,
}) => {
  const pageSchema = {
    '@context': 'http://schema.org',
    '@type': schemaType,
    headline,
    url,
    description,
    primaryImageOfPage: buildImage(image),
    author: {
      '@id': '#person',
    },
    copyrightHolder: {
      '@id': '#person',
    },
  }

  return JSON.stringify([personSchema({ name, jobTitle, sameAs }), pageSchema])
}

const articleSchema = ({
  headline,
  url,
  description,
  image,
  fallbackImage,
  datePublished,
  dateModified,
  name,
  siteName,
  jobTitle,
  sameAs,
}) => {
  const articleSchema = {
    '@context': 'http://schema.org',
    '@type': 'BlogPosting',
    headline,
    url,
    description,
    image: buildImage(image),
    author: {
      '@id': '#person',
    },
    datePublished,
    dateModified,
    publisher: {
      '@id': '#publisher',
    },
    mainEntityOfPage: {
      '@type': 'WebPage',
      '@id': url,
    },
  }

  return JSON.stringify([
    personSchema({ name, jobTitle, sameAs }),
    publisherSchema({ siteName, logo: fallbackImage }),
    articleSchema,
  ])
}

export default function (data) {
  const isPage = data.templateKey.indexOf('-page') !== -1
  const isArticle = data.templateKey.indexOf('-post') !== -1
  const isArchive = data.templateKey.indexOf('-archive') !== -1

  if (isPage || isArchive) {
    return webpageSchema(data)
  } else if (isArticle) {
    return articleSchema(data)
  }
}
