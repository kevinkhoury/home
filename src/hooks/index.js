import useNavPages from './useNavPages'
import useSiteData from './useSiteData'
import useStructuredData from './useStructuredData'
import useSiteMeta from './useSiteMeta'
import useRecentPosts from './useRecentPosts'

export {
  useNavPages,
  useSiteData,
  useStructuredData,
  useSiteMeta,
  useRecentPosts,
}
