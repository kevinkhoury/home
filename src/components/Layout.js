import React, { Fragment, useState, useContext } from 'react'
import PropTypes from 'prop-types'
import Nav from './Nav'
import Fonts from './Fonts'
import { Link } from 'gatsby'
import ThemeSwitcher from './ThemeSwitcher'
import { ThemeOptionsContext } from '../context/ThemeOptions'
import { useSiteData } from '../hooks'
import SEO, { seoPropTypes } from './SEO'

import '../style/all.sass'

const Layout = ({ seoProps, children }) => {
  const [toggleNav, setToggleNav] = useState(false)
  const {
    colorScheme,
    setColorScheme,
    fontScheme,
    setFontScheme,
    previewOpen,
    setPreviewOpen,
  } = useContext(ThemeOptionsContext)
  const {
    name,
    siteName,
    themeOptions: { showThemeSwitcher, fontScheme: ssrFontScheme },
  } = useSiteData()

  return (
    <Fragment>
      <Fonts
        fontScheme={typeof window === 'undefined' ? ssrFontScheme : fontScheme}
      />
      <SEO {...seoProps} />
      <div className={`site-wrapper ${toggleNav ? `site-head-open` : ``}`}>
        <header className="site-head">
          <div className="site-head-container">
            <Nav
              toggleNav={toggleNav}
              setToggleNav={setToggleNav}
              siteName={siteName}
            />
            {!!showThemeSwitcher && (
              <ThemeSwitcher
                colorScheme={colorScheme}
                setColorScheme={setColorScheme}
                fontScheme={fontScheme}
                setFontScheme={setFontScheme}
                previewOpen={previewOpen}
                setPreviewOpen={setPreviewOpen}
              />
            )}
          </div>
        </header>

        <main id="site-main" className="site-main">
          <div id="swup" className="transition-fade">
            {children}
          </div>
        </main>

        <footer className="site-foot">
          &copy; {new Date().getFullYear()} <Link to={`/`}>{name}</Link>, all
          rights reserved.
        </footer>
      </div>
    </Fragment>
  )
}

Layout.propTypes = {
  seoProps: PropTypes.shape(seoPropTypes).isRequired,
  children: PropTypes.node.isRequired,
}

export default Layout
