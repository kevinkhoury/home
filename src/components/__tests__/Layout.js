import React, {
  useState as useStateMock,
  useContext as useContextMock,
} from 'react'
import { useSiteData as useSiteDataMock } from '../../hooks'
import moment from 'moment'
import MockDate from 'mockdate'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import Layout from '../Layout'

const DEFAULT_PROPS = {
  seoProps: {
    pageTitle: 'pagetitle',
    metaDescription: 'metadescription',
    featuredImage: {
      src: {
        childImageSharp: {
          fluid: {},
          original: {
            height: 100,
            width: 200,
          },
        },
      },
      d: {
        childImageSharp: {},
      },
      m: {
        childImageSharp: {},
      },
      alt: 'imgalt',
      caption: 'imgcaption',
    },
    slug: '/foobar',
    date: moment('20200101', 'YYYYMMDD'),
    dateModified: moment('20200101', 'YYYYMMDD'),
    templateKey: 'templatekey',
    schemaType: 'schematype',
  },
  children: <div>children</div>,
}

jest.mock('../SEO', () => () => 'SEO')
jest.mock('../Fonts', () => () => 'Fonts')
jest.mock('../Nav', () => () => 'Nav')
jest.mock('../ThemeSwitcher', () => () => 'ThemeSwitcher')
jest.mock('../../context/ThemeOptions', () => () => 'ThemeOptionsContext')
jest.mock('../../hooks', () => ({
  useSiteData: jest.fn(),
}))
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useContext: jest.fn(),
}))

beforeEach(() => {
  MockDate.set('2020-01-01')
  useStateMock.mockImplementation((init) => [init, jest.fn()])
  useContextMock.mockImplementation(() => ({
    colorScheme: 'colorScheme',
    setColorScheme: jest.fn(),
    fontScheme: 'fontScheme',
    setFontScheme: jest.fn(),
    previewOpen: true,
    setPreviewOpen: jest.fn(),
  }))
  useSiteDataMock.mockImplementation(() => ({
    name: 'name',
    siteName: 'siteName',
    themeOptions: {
      showThemeSwitcher: true,
      fontScheme: 'fontScheme',
    },
  }))
})

afterEach(() => {
  MockDate.reset()
  jest.clearAllMocks()
})

describe('Layout', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<Layout {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('shows theme switcher with default context mock', () => {
    render(<Layout {...DEFAULT_PROPS} />)
    const result = screen.getByText(/ThemeSwitcher/)
    expect(result).toBeTruthy()
  })
})
