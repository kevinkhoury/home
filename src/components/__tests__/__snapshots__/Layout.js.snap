// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout snapshot renders properly 1`] = `
Array [
  "Fonts",
  "SEO",
  <div
    className="site-wrapper "
  >
    <header
      className="site-head"
    >
      <div
        className="site-head-container"
      >
        Nav
        ThemeSwitcher
      </div>
    </header>
    <main
      className="site-main"
      id="site-main"
    >
      <div
        className="transition-fade"
        id="swup"
      >
        <div>
          children
        </div>
      </div>
    </main>
    <footer
      className="site-foot"
    >
      Â© 
      2019
       
      <a
        href="/"
      >
        name
      </a>
      , all rights reserved.
    </footer>
  </div>,
]
`;
